### PR TITLE
Load Applications via Gems instead of Source Tree

### PR DESCRIPTION
**Pull request by danielsdeleo**
This fixes an issue where files in the application could be multiply-required when the application is also installed as a gem, and the application loads a plugin which has a rubygems runtime dependency on the main application. In particular, ruby applications are installed as gems in Chef/ChefDK omnibus packages in addition to the source checkouts, and chef-provisioning-aws (among others) is a Chef plugin which specifies a runtime dependency on "chef" in its gemspec. Due to a change in Rubygems 2.4.8 (probably introduced in 2.4.5, but this hasn't been verified, I suspect it is this: https://github.com/rubygems/rubygems/commit/e2935c85fc62594ccd567c8facd46ddbd01ad100), after loading the plugin, requires of files in the "chef" application would resolve to the gem install rather than the source checkout, causing double requires which result in a bunch of redefined constant warnings, and finally a SuperclassMismatch for code like:

```ruby
# If you run this twice, you get a different `Struct.new(:item)` class, resulting in an error
class Foo < Struct.new(:item)
end
``` 

To resolve the issue, we now use a more rubygems-centric loading mechanism, where we look up the executable via the Gem::Specification and then load it therefore loading only the rubygems copy of the application. This does leave us in a kinda ugly state because you need the source tree to generate the appbundler executables, but at runtime only the rubygems code is loaded. This could be improved quite a bit, but would likely require some changes to our Omnibus build definitions, so that will be left for a later time.

@jkeiser @ksubrama

### CODE REVIEW

File: lib/appbundler/app.rb [hunk 1]
```diff
@@ -116,14 +116,15 @@ def binstub(bin_file)
...
 
     def load_statement_for(bin_file)
+      name, version = app_spec.name, app_spec.version
+      bin_basename = File.basename(bin_file)
       <<-E
-bin_dir = File.dirname(__FILE__)
-if File.symlink?(__FILE__)
-  bin_dir = File.dirname(File.readlink(__FILE__))
-end
+gem "#{name}", "= #{version}"
```
Comment (jkeiser) [inline lib/appbundler/app.rb]: Is this one needed? Shouldn't it be written out in the massive bunch of `gem` commands that pin the world?

**Comment by lamont-granquist** [comment]
Well ruby 1.9.3 and ohai 8.5.1 got into a fight and your travis build lost.

So this means that the git checkout in /opt/chef/embedded/apps/chef will no longer be hackable without doing a rake install or whatever?

File: lib/appbundler/app.rb (hunk 1 above)
Comment (lamont-granquist) [inline lib/appbundler/app.rb]: I think the rest of the world is handled here:

https://github.com/chef/appbundler/blob/use-installed-gems/lib/appbundler/app.rb#L106

Comment (jaym) [inline lib/appbundler/app.rb]: It looks like it is not:

```ruby
#!/Users/jmundrawala/.rbenv/versions/2.1.2/bin/ruby
#--APP_BUNDLER_BINSTUB_FORMAT_VERSION=1--
ENV["GEM_HOME"] = ENV["GEM_PATH"] = nil unless ENV["APPBUNDLER_ALLOW_RVM"] == "true"
gem "chef-config", "= 12.5.0.current.0"
gem "mixlib-config", "= 2.2.0"
gem "mixlib-shellout", "= 2.2.0"
gem "chef-zero", "= 4.2.3"
gem "ffi-yajl", "= 2.2.2"
gem "libyajl2", "= 1.2.0"
gem "hashie", "= 2.1.2"
gem "mixlib-log", "= 1.6.0"
gem "rack", "= 1.5.5"
gem "uuidtools", "= 2.1.5"
gem "diff-lcs", "= 1.2.5"
gem "erubis", "= 2.7.0"
gem "highline", "= 1.7.3"
gem "mixlib-authentication", "= 1.3.0"
gem "mixlib-cli", "= 1.5.0"
gem "net-ssh", "= 2.9.2"
gem "net-ssh-multi", "= 1.2.1"
gem "net-ssh-gateway", "= 1.2.0"
gem "ohai", "= 8.6.0"
gem "ffi", "= 1.9.10"
gem "ipaddress", "= 0.8.0"
gem "mime-types", "= 2.6.1"
gem "rake", "= 10.4.2"
gem "systemu", "= 2.6.5"
gem "wmi-lite", "= 1.0.0"
gem "plist", "= 3.1.0"
gem "pry", "= 0.10.1"
gem "coderay", "= 1.1.0"
gem "method_source", "= 0.8.2"
gem "slop", "= 3.6.0"
gem "rspec-core", "= 3.3.2"
gem "rspec-support", "= 3.3.0"
gem "rspec-expectations", "= 3.3.1"
gem "rspec-mocks", "= 3.3.2"
gem "rspec_junit_formatter", "= 0.2.3"
gem "builder", "= 3.2.2"
gem "serverspec", "= 2.23.0"
gem "multi_json", "= 1.11.2"
gem "rspec", "= 3.3.0"
gem "rspec-its", "= 1.2.0"
gem "specinfra", "= 2.43.0"
gem "net-scp", "= 1.2.1"
gem "net-telnet", "= 0.1.1"
gem "sfl", "= 2.2"
gem "syslog-logger", "= 1.6.8"
gem "chef", "= 12.5.0.current.0"

spec = Gem::Specification.find_by_name("chef", "= 12.5.0.current.0")
bin_file = spec.bin_file("chef-client")

Kernel.load(bin_file)
```

**Comment by danielsdeleo** [comment]
Yeah, to hack the code, you can either just edit in the gem dir, or edit the copy in embedded/apps and rake install it. /cc @fnichol might break your updater script :crying_cat_face:

File: lib/appbundler/app.rb (hunk 1 above)
Comment (danielsdeleo) [inline lib/appbundler/app.rb]: Yeah, the other statements come from traversing the runtime deps, so you won't have the top-level app in there.

**Comment by lamont-granquist** [comment]
Yeah that should probably be socialized somehow, @jaymzh also does it by hand sometimes.

**Comment by jaym** [comment]
This LGTM. 1 question: Do we actually need anything in the apps directory after this?

**Comment by danielsdeleo** [comment]
@jaym it may be possible to put just a Gemfile in there, like:

```ruby
# /opt/chefdk/embedded/apps/chef/Gemfile
gem "chef"
```

Then bundle install that. But of course you would have to tweak that to get the right commit and such.

**Comment by lamont-granquist** [comment]
I actually sort of like this direction...  You can just hack up the gem library now which is right where everyone expects to be able to hack it up.  The only gotcha is that if you update gems and bundle install that the pins in the binary won't update.  It feels a bit more 'ruby' than what we were doing before...

**Comment by danielsdeleo** [comment]
Yeah, I think the intermediate state we're going to be in when we merge this is weird (hence why I call it out), but if we continue making progress in this direction it could be good. In the meantime, we can roll forward on rubygems, which is a plus.

**Comment by lamont-granquist** [comment]
Yep, :+1: on the direction, good enough for today.

**Comment by ksubrama** [comment]
:+1:

**Pull request merged**